epoxy-paging3/src/main/java/com/airbnb/epoxy/paging3/PagedDataModelCache.kt

```diff
@@ -0,0 +1,198 @@
+     * that happens.
+     */
+    private fun assertUpdateCallbacksAllowed() {
+        require(inSubmitList || Looper.myLooper() == modelBuildingHandler.looper) {
+            "The notify executor for your PagedList must use the same thread as the model building handler set in PagedListEpoxyController.modelBuildingHandler"
+        }
+    }
+
+    private val dispatcher = modelBuildingHandler.asCoroutineDispatcher()
+    private val asyncDiffer = AsyncPagingDataDiffer(
+        diffCallback = itemDiffCallback,
+        updateCallback = updateCallback,
+        mainDispatcher = dispatcher,
+        workerDispatcher = dispatcher
+    )
+
+    @Synchronized
+    suspend fun submitData(pagingData: PagingData<T>) {
+        inSubmitList = true
+        asyncDiffer.submitData(pagingData)
+        inSubmitList = false
+    }
+
+    @Synchronized
+    fun getModels(): List<EpoxyModel<*>> {
+        val currentList = asyncDiffer.snapshot()
+
+        // The first time models are built the EpoxyController does so synchronously, so that
+        // the UI can be ready immediately. To avoid concurrent modification issues with the PagedList
+        // and model cache we can't allow that first build to touch the cache.
+        if (Looper.myLooper() != modelBuildingHandler.looper) {
+            return currentList.mapIndexed { position, item ->
+                modelBuilder(position, item)
+            }
+        }
+
+        (0 until modelCache.size).forEach { position ->
+            if (modelCache[position] == null) {
+                modelCache[position] = modelBuilder(position, currentList[position])
+            }
+        }
+
+        lastPosition?.let {
+            triggerLoadAround(it)
+        }
+        @Suppress("UNCHECKED_CAST")
+        return modelCache as List<EpoxyModel<*>>
+    }
+
+    /**
+     * Clears all cached models to force them to be rebuilt next time models are retrieved.
+     * This is posted to the model building thread to maintain data synchronicity.
+     */
+    fun clearModels() {
+        modelBuildingHandler.post {
+            clearModelsSynchronized()
+        }
+    }
+
+    fun retry() {
+        asyncDiffer.retry()
+    }
+
+    fun refresh() {
+        asyncDiffer.refresh()
+    }
+
+    fun addLoadStateListener(listener: (CombinedLoadStates) -> Unit) {
+        asyncDiffer.addLoadStateListener(listener)
+    }
+
+    fun removeLoadStateListener(listener: (CombinedLoadStates) -> Unit) {
+        asyncDiffer.removeLoadStateListener(listener)
+    }
+
+    @Synchronized
+    private fun clearModelsSynchronized() {
+        modelCache.fill(null)
+    }
+
+    fun loadAround(position: Int) {
+        triggerLoadAround(position)
+        lastPosition = position
+    }
+
+    private fun triggerLoadAround(position: Int) {
+        if (asyncDiffer.itemCount > 0) {
+            asyncDiffer.getItem(position.coerceAtMost(asyncDiffer.itemCount - 1))
```

Comment:
Shouldn't here be `coerceIn`?
```suggestion
            asyncDiffer.getItem(position.coerceIn(0, asyncDiffer.itemCount - 1))
```

Reply:
Isn't that we have already provided a lower bound with `if (asyncDiffer.itemCount > 0)`?

Reply:
Is it bounds `position`?

Reply:
oh I see, I misunderstood your point, will adapt this change. Thanks 👍